src/lib/events.ts

```diff
@@ -0,0 +1,105 @@
+import { Promise } from "bluebird";
+import { BaseContract, Contract, ContractEventName, EventLog } from "ethers";
+
+const defaultConcurrency = 200;
+const maxRetries = 3;
+const retrySleepTime = 10;
+
+function delay(s: number) {
+  return new Promise((resolve) => setTimeout(resolve, s * 1000));
+}
+export interface EventSearchConfig {
+  fromBlock: number;
+  toBlock: number;
+  maxBlockLookBack?: number;
+  concurrency?: number;
+}
+
+export async function paginatedEventQuery<T extends EventLog>(
+  contract: Contract | BaseContract,
+  filter: ContractEventName,
+  searchConfig: EventSearchConfig,
+  retryCount = 0
+): Promise<Array<T>> {
+  // If the max block look back is set to 0 then we dont need to do any pagination and can query over the whole range.
+  if (searchConfig.maxBlockLookBack === 0)
+    return (await contract.queryFilter(filter, searchConfig.fromBlock, searchConfig.toBlock)) as Array<T>;
+
+  // Compute the number of queries needed. If there is no maxBlockLookBack set then we can execute the whole query in
+  // one go. Else, the number of queries is the range over which we are searching, divided by the maxBlockLookBack,
+  // rounded up. This gives us the number of queries we need to execute to traverse the whole block range.
+  const paginatedRanges = getPaginatedBlockRanges(searchConfig);
+
+  try {
+    return (
+      (
+        await Promise.map(paginatedRanges, ([fromBlock, toBlock]) => contract.queryFilter(filter, fromBlock, toBlock), {
+          concurrency: typeof searchConfig.concurrency == "number" ? searchConfig.concurrency : defaultConcurrency,
+        })
+      )
+        .flat()
```

Comment:
did you need to use the bluebird promise due to limitations of the number of requests, that would error/fail without pagination?

Reply:
I reused this existing code, used in other parts of our codebase without modifications to minimize impact. Additionally, the pagination that this has by default, I think, makes our usage of the node URLs efficient without adding much delay. Would you change it to normal Promises?